From 10b0a82884c98e4a0d7002079eaf7601b536e9bb Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Thu, 24 Mar 2022 20:46:07 +0800
Subject: [PATCH] config: add local mkdir_p implementation

---
 src/config.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

--- a/src/config.c
+++ b/src/config.c
@@ -310,6 +310,31 @@ static const char *svc_param_key_names[D
 	[DNR_SVC_OHTTP] = "ohttp",
 };
 
+static int mkdir_p(char *dir, mode_t mask)
+{
+	char *l = strrchr(dir, '/');
+	int ret;
+
+	if (!l)
+		return 0;
+
+	*l = '\0';
+
+	if (mkdir_p(dir, mask))
+		return -1;
+
+	*l = '/';
+
+	ret = mkdir(dir, mask);
+	if (ret && errno == EEXIST)
+		return 0;
+
+	if (ret)
+		syslog(LOG_ERR, "mkdir(%s, %d) failed: %m\n", dir, mask);
+
+	return ret;
+}
+
 static void set_interface_defaults(struct interface *iface)
 {
 	iface->ignore = true;
