From f85c33fdf4f0ba156f0fec790ba93b2dee1c6aaf Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Fri, 26 Nov 2021 18:39:09 +0800
Subject: [PATCH] mac80211: do not access the IV when it was stripped

---
 net/mac80211/rx.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -1831,7 +1831,7 @@ ieee80211_rx_h_decrypt(struct ieee80211_
 	if (rx->sta) {
 		int keyid = rx->sta->ptk_idx;
 
-		if (ieee80211_has_protected(fc) && rx->sta->cipher_scheme) {
+		if (ieee80211_has_protected(fc) && rx->sta->cipher_scheme && !(status->flag & RX_FLAG_IV_STRIPPED)) {
 			cs = rx->sta->cipher_scheme;
 			keyid = ieee80211_get_cs_keyid(cs, rx->skb);
 			if (unlikely(keyid < 0))
