From 3e50231511cae24dbddebd4aa18d92b9f0c28e16 Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Mon, 7 Nov 2022 11:52:21 +0800
Subject: [PATCH 3/6] fix uninitialized value

---
 network.cpp | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/network.cpp b/network.cpp
index 5699c9c..2467bcc 100644
--- a/network.cpp
+++ b/network.cpp
@@ -780,7 +780,7 @@ void init_filter(int port)
 		}
 	}
 
-	int dummy;
+	int dummy = 0;
 
 	int ret=setsockopt(raw_recv_fd, SOL_SOCKET, SO_DETACH_FILTER, &dummy, sizeof(dummy)); //in case i forgot to remove
 	if (ret != 0)
@@ -928,7 +928,7 @@ void init_filter(int port)
 		bpf.filter = code_icmp;
 	}
 
-	int dummy;
+	int dummy = 0;
 
 	int ret=setsockopt(raw_recv_fd, SOL_SOCKET, SO_DETACH_FILTER, &dummy, sizeof(dummy)); //in case i forgot to remove
 	if (ret != 0)
@@ -951,7 +951,7 @@ void remove_filter()
 {
 	filter_port=0;
 #ifdef UDP2RAW_LINUX
-	int dummy;
+	int dummy = 0;
 	int ret=setsockopt(raw_recv_fd, SOL_SOCKET, SO_DETACH_FILTER, &dummy, sizeof(dummy));
 	if (ret != 0)
 	{
-- 
2.25.1

