From ec36a9a475d965f346826df722dfee34de1c7f55 Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Sun, 23 Jul 2023 22:41:51 +0800
Subject: [PATCH] fix max_addr_len

---
 common.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/common.h
+++ b/common.h
@@ -159,7 +159,7 @@ typedef u64_t anti_replay_seq_t;
 
 typedef u64_t my_time_t;
 
-const int max_addr_len = 100;
+const int max_addr_len = 300;
 
 extern int force_socket_buf;
 
