From d22efe9b29faffaa8cbb3c2ab29c222482e8928b Mon Sep 17 00:00:00 2001
From: HiGarfield <32226909+HiGarfield@users.noreply.github.com>
Date: Sat, 14 Feb 2026 17:55:43 +0800
Subject: [PATCH] Fix string length overflow with a long user password

---
 encrypt.cpp | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- a/encrypt.cpp
+++ b/encrypt.cpp
@@ -53,14 +53,14 @@ int aes128cfb_old = 0;
 // TODO key negotiation and forward secrecy
 
 int my_init_keys(const char *user_passwd, int is_client) {
-    char tmp[1000] = "";
+    string tmp;
     int len = strlen(user_passwd);
 
-    strcat(tmp, user_passwd);
+    tmp.reserve(len + 4);
+    tmp.append(user_passwd);
+    tmp.append("key1");
 
-    strcat(tmp, "key1");
-
-    md5((uint8_t *)tmp, strlen(tmp), (uint8_t *)normal_key);
+    md5((uint8_t *)tmp.data(), tmp.size(), (uint8_t *)normal_key);
 
     if (auth_mode == auth_hmac_sha1)
         is_hmac_used = 1;
