diff --git a/net/core/dev.c b/net/core/dev.c
index f34aad63..49c8f415 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3967,7 +3967,16 @@ another_round:
 		goto ncls;
 	}
 #endif
+#ifdef CONFIG_NET_INGRESS
+	if (static_key_false(&ingress_needed)) {
+		skb = handle_ing(skb, &pt_prev, &ret, orig_dev);
+		if (!skb)
+			goto out;
 
+		if (nf_ingress(skb, &pt_prev, &ret, orig_dev) < 0)
+			goto out;
+	}
+#endif
 	if (pfmemalloc)
 		goto skip_taps;
 
@@ -3984,16 +3993,6 @@ another_round:
 	}
 
 skip_taps:
-#ifdef CONFIG_NET_INGRESS
-	if (static_key_false(&ingress_needed)) {
-		skb = handle_ing(skb, &pt_prev, &ret, orig_dev);
-		if (!skb)
-			goto out;
-
-		if (nf_ingress(skb, &pt_prev, &ret, orig_dev) < 0)
-			goto out;
-	}
-#endif
 #ifdef CONFIG_NET_CLS_ACT
 	skb->tc_verd = 0;
 ncls:
