From aed6d9315eb76f8e5d68c5fcc77df1afee75b8f5 Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Tue, 23 May 2023 23:10:05 +0800
Subject: [PATCH] arch/arm/include/asm/uaccess.h: fix includes

---
 arch/arm/include/asm/uaccess.h | 1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm/include/asm/uaccess.h
+++ b/arch/arm/include/asm/uaccess.h
@@ -11,6 +11,7 @@
 /*
  * User space memory access functions
  */
+#include <linux/kernel.h>
 #include <linux/string.h>
 #include <linux/thread_info.h>
 #include <asm/errno.h>
