--- a/print.c
+++ b/print.c
@@ -148,6 +148,7 @@ size_t *sz; /* returned remaining size */
  */
 
 static void fill_portmap() {
+#if !defined __UCLIBC__ || (defined __UCLIBC__ && defined __UCLIBC_HAS_RPC__)
     static int already_run = 0;
     char buf[128], *cp, *nm;
     CLIENT *c;
@@ -273,6 +274,7 @@ static void fill_portmap() {
         Pth[pr][h] = pt;
     }
     clnt_destroy(c);
+#endif
 }
 #endif /* !defined(HASNORPC_H) */
 
