From e518e87533345f53fb59e1b9e99994dd73eb8942 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Fri, 9 Sep 2022 15:56:54 +0100
Subject: [PATCH 05/39] Fix namebuff overwrite leading to wrong log after
 socket bind warning.

---
 src/forward.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/forward.c b/src/forward.c
index aa9ace0..9d1f005 100644
--- a/src/forward.c
+++ b/src/forward.c
@@ -2439,12 +2439,12 @@ static int random_sock(struct server *s)
 	return fd;
 
       if (s->interface[0] == 0)
-	(void)prettyprint_addr(&s->source_addr, daemon->namebuff);
+	(void)prettyprint_addr(&s->source_addr, daemon->addrbuff);
       else
-	strcpy(daemon->namebuff, s->interface);
+	safe_strncpy(daemon->addrbuff, s->interface, ADDRSTRLEN);
 
       my_syslog(LOG_ERR, _("failed to bind server socket to %s: %s"),
-		daemon->namebuff, strerror(errno));
+		daemon->addrbuff, strerror(errno));
       close(fd);
     }
   
-- 
2.25.1

