From f32498465dfb1f885c83b52576f5afa6f701e211 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Fri, 16 Sep 2022 09:35:44 +0100
Subject: [PATCH 20/31] Don't exclude stale-cache answers from "local answered"
 metric.

---
 src/forward.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/forward.c b/src/forward.c
index 14d9421..45f0176 100644
--- a/src/forward.c
+++ b/src/forward.c
@@ -1820,7 +1820,9 @@ void receive_query(struct listener *listen, time_t now)
 #endif
 	  send_from(listen->fd, option_bool(OPT_NOWILD) || option_bool(OPT_CLEVERBIND),
 		    (char *)header, m, &source_addr, &dst_addr, if_index);
-	  daemon->metrics[stale ? METRIC_DNS_STALE_ANSWERED : METRIC_DNS_LOCAL_ANSWERED]++;
+	  daemon->metrics[METRIC_DNS_LOCAL_ANSWERED]++;
+	  if (stale)
+	    daemon->metrics[METRIC_DNS_STALE_ANSWERED]++;
 	}
       
       if (m == 0 || stale)
-- 
2.25.1

