diff --git a/luci/sqm-cbi.lua b/luci/sqm-cbi.lua
index 56db8fd..6307651 100644
--- a/luci/sqm-cbi.lua
+++ b/luci/sqm-cbi.lua
@@ -61,7 +61,13 @@ n = s:taboption("tab_basic", ListValue, "interface", translate("Interface name")
 -- sm lifted from luci-app-wol, the original implementation failed to show pppoe-ge00 type interface names
 for _, iface in ipairs(ifaces) do
 	if not (iface == "lo" or iface:match("^ifb.*")) then
-		n:value(iface)
+		local nets = net:get_interface(iface)
+		nets = nets and nets:get_networks() or {}
+		for k, v in pairs(nets) do
+			nets[k] = nets[k].sid
+		end
+		nets = table.concat(nets, ",")
+		n:value(iface, ((#nets > 0) and "%s (%s)" % {iface, nets} or iface))
 	end
 end
 n.rmempty = false
@@ -75,7 +81,7 @@ ul = s:taboption("tab_basic", Value, "upload", translate("Upload speed (kbit/s)
 ul.datatype = "and(uinteger,min(0))"
 ul.rmempty = false
 
-dbl = s:taboption("tab_basic", Flag, "debug_logging", translate("Create log file for this SQM instance under /var/run/sqm/${Inerface_name}.debug.log. Make sure to delete log files manually."))
+dbl = s:taboption("tab_basic", Flag, "debug_logging", translate("Create log file for this SQM instance under /var/run/sqm/${Interface_name}.debug.log. Make sure to delete log files manually."))
 dbl.rmempty = false
 
 
