From 60365ae82bb0ffca93b49fcbf736b33e9af2b15d Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Thu, 27 Mar 2025 00:05:02 +0800
Subject: [PATCH] dns_client: fix compatibility with openssl 1.0

---
 src/dns_client.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/src/dns_client.c
+++ b/src/dns_client.c
@@ -332,7 +332,12 @@ static ssize_t _ssl_write_ext2(struct dn
 #ifdef OSSL_QUIC1_VERSION
 	ret = SSL_write_ex2(ssl, buff, num, flags, &written);
 #else
+#if OPENSSL_VERSION_NUMBER >= 0x10101000L && !defined(OPENSSL_NO_SSL_WRITE_EX)
 	ret = SSL_write_ex(ssl, buff, num, &written);
+#else
+	ret = SSL_write(ssl, buff, num);
+	written = (ret > 0) ? (size_t)ret : 0;
+#endif
 #endif
 	pthread_mutex_unlock(&server->lock);
 
