From 9469015ebf0b7d63cdc12e61084fcc88c94a618b Mon Sep 17 00:00:00 2001
From: Nick Peng <pymumu@gmail.com>
Date: Sat, 17 May 2025 23:41:42 +0800
Subject: [PATCH] tlog: update tlog

---
 src/tlog.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/src/tlog.c
+++ b/src/tlog.c
@@ -761,7 +761,7 @@ int tlog_vext(tlog_level level, const ch
         return _tlog_early_print(&info_inter, format, ap);
     }
 
-    if (unlikely(tlog.root->logsize <= 0)) {
+    if (unlikely(tlog.root->logsize <= 0 && tlog.root->logscreen == 0 && tlog.root->set_custom_output_func == 0)) {
         return 0;
     }
 
@@ -1269,7 +1269,7 @@ static int _tlog_write(struct tlog_log *
         _tlog_archive_log(log);
     }
 
-    if (log->fd <= 0) {
+    if (log->fd <= 0 && log->logsize > 0) {
         /* open a new log file to write */
         time_t now;
 
