From 98a6403ce60254c44a5a7253ad1df3d2635f3b0d Mon Sep 17 00:00:00 2001
From: Nick Peng <pymumu@gmail.com>
Date: Fri, 12 Dec 2025 21:12:58 +0800
Subject: [PATCH] fix: avoid ping block caused by a large number of requests

---
 src/fast_ping.c | 23 +++++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

--- a/src/fast_ping.c
+++ b/src/fast_ping.c
@@ -749,8 +749,22 @@ static int _fast_ping_sendping_v6(struct
 				 ping_host->addr_len);
 	if (len != sizeof(struct fast_ping_packet)) {
 		int err = errno;
-		if (errno == ENETUNREACH || errno == EINVAL || errno == EADDRNOTAVAIL || errno == EHOSTUNREACH) {
+		switch (err) {
+		case ENETUNREACH:
+		case EINVAL:
+		case EADDRNOTAVAIL:
+		case EHOSTUNREACH:
+		case ENOBUFS:
+		case EACCES:
+		case EPERM:
+		case EAFNOSUPPORT:
+		case EAGAIN:
+#if EWOULDBLOCK != EAGAIN
+		case EWOULDBLOCK:
+#endif
 			goto errout;
+		default:
+			break;
 		}
 
 		if (is_private_addr_sockaddr(&ping_host->addr, ping_host->addr_len)) {
@@ -841,6 +855,9 @@ static int _fast_ping_sendping_v4(struct
 	len = sendto(ping.fd_icmp, packet, sizeof(struct fast_ping_packet), 0, &ping_host->addr, ping_host->addr_len);
 	if (len != sizeof(struct fast_ping_packet)) {
 		int err = errno;
+		if (errno == EAGAIN || errno == EWOULDBLOCK) {
+			goto errout;
+		}
 		if (errno == ENETUNREACH || errno == EINVAL || errno == EADDRNOTAVAIL || errno == EPERM || errno == EACCES) {
 			goto errout;
 		}
@@ -1018,7 +1035,8 @@ static int _fast_ping_create_icmp_sock(F
 	int fd = -1;
 	struct ping_host_struct *icmp_host = NULL;
 	struct epoll_event event;
-	int buffsize = 64 * 1024;
+	/* Set receive and send buffer to 512KB， if buffer size is too small, ping may fail. */
+	int buffsize = 512 * 1024;
 	socklen_t optlen = sizeof(buffsize);
 	const int val = 255;
 	const int on = 1;
@@ -1083,6 +1101,7 @@ static int _fast_ping_create_icmp_sock(F
 	setsockopt(fd, SOL_SOCKET, SO_RCVBUF, (const char *)&buffsize, optlen);
 	setsockopt(fd, SOL_IP, IP_TTL, &val, sizeof(val));
 	setsockopt(fd, IPPROTO_IP, IP_TOS, &ip_tos, sizeof(ip_tos));
+	set_fd_nonblock(fd, 1);
 
 	icmp_host->fd = fd;
 	icmp_host->type = type;
