From 365e78aa6bd89eba5798e57c6e6cfbce04aac4ed Mon Sep 17 00:00:00 2001
From: Nick Peng <pymumu@gmail.com>
Date: Mon, 22 Dec 2025 22:42:53 +0800
Subject: [PATCH] dns_cache: fix crash caused by race condition

---
 src/dns_cache.c | 9 +++++++++
 1 file changed, 9 insertions(+)

--- a/src/dns_cache.c
+++ b/src/dns_cache.c
@@ -139,6 +139,15 @@ void dns_cache_release(struct dns_cache
 
 static void _dns_cache_remove(struct dns_cache *dns_cache)
 {
+	/*
+	 * If the list is empty, it means the cache has been removed by other threads.
+	 * The reference count has been decreased by the other thread.
+	 * We should simply return to avoid double free.
+	 */
+	if (list_empty(&dns_cache->list)) {
+		return;
+	}
+
 	hash_del(&dns_cache->node);
 	list_del_init(&dns_cache->list);
 	if (dns_timer_del(&dns_cache->timer)) {
