From c1a275a836b41cc0cc839e6501ca9568dc4bc8c8 Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Thu, 2 Apr 2026 00:56:11 +0800
Subject: [PATCH 2/2] try to fix client MAC NULL issue

---
 src/util.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/src/util.c
+++ b/src/util.c
@@ -1070,7 +1070,7 @@ int netlink_get_neighbors(int family, co
 	nlh = (struct nlmsghdr *)buf;
 	nlh->nlmsg_len = NLMSG_LENGTH(sizeof(struct ndmsg));
 	nlh->nlmsg_type = RTM_GETNEIGH;
-	nlh->nlmsg_flags = NLM_F_REQUEST;
+	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
 	nlh->nlmsg_seq = time(NULL);
 	nlh->nlmsg_pid = getpid();
 
@@ -1092,7 +1092,7 @@ int netlink_get_neighbors(int family, co
 		}
 
 		send_count++;
-		if (send(netlink_neighbor_fd, buf, NLMSG_SPACE(sizeof(struct ndmsg)), 0) < 0) {
+		if (send(netlink_neighbor_fd, buf, NLMSG_ALIGN(nlh->nlmsg_len), 0) < 0) {
 			if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
 				struct timespec waiter;
 				waiter.tv_sec = 0;
