From 518ad70bd1340de7e63ae806524bfb1e5350959d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Mat=C4=9Bjek?= <martin.matejek@gmx.com>
Date: Sun, 23 Oct 2022 22:05:48 +0200
Subject: [PATCH] datastorage: fix use-after-free compiler error
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Closes: #198

Signed-off-by: Martin Matějek <martin.matejek@gmx.com>
---
 src/storage/datastorage.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/storage/datastorage.c
+++ b/src/storage/datastorage.c
@@ -1332,8 +1332,8 @@ void insert_macs_from_file() {
         }
     }
 
-    fclose(fp);
     dawn_unregmem(fp);
+    fclose(fp);
     if (line)
     {
         free(line);
