From 1806e0fbc6f32d5db3c1efc357f870ad24c0e064 Mon Sep 17 00:00:00 2001
From: HarvsG <11440490+HarvsG@users.noreply.github.com>
Date: Fri, 15 Sep 2023 11:00:05 +0100
Subject: [PATCH] don't kick if data rate is at threshold

---
 src/storage/datastorage.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/storage/datastorage.c
+++ b/src/storage/datastorage.c
@@ -565,7 +565,7 @@ int kick_clients(struct dawn_mac bssid_m
                     // only use rx_rate for indicating if transmission is going on
                     // <= 6MBits <- probably no transmission
                     // tx_rate has always some weird value so don't use ist
-                    if (have_bandwidth_iwinfo && dawn_metric.bandwidth_threshold != 0 && rx_rate > dawn_metric.bandwidth_threshold) {
+                    if (have_bandwidth_iwinfo && dawn_metric.bandwidth_threshold != 0 && rx_rate >= dawn_metric.bandwidth_threshold) {
                         dawnlog_info("Client " MACSTR ": Don't kick due to active data transfer: RX rate %f exceeds %d limit\n", MAC2STR(j->client_addr.u8), rx_rate, dawn_metric.bandwidth_threshold);
                     }
                     else
