--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -48,7 +48,8 @@
 
 ADD_EXECUTABLE(dawn ${SOURCES})
 
-TARGET_LINK_LIBRARIES(dawn ${LIBS})
+find_package(Threads)
+TARGET_LINK_LIBRARIES(dawn ${LIBS} ${CMAKE_THREAD_LIBS_INIT})
 
 INSTALL(TARGETS dawn
         RUNTIME DESTINATION /usr/sbin/)
--- a/src/network/networksocket.c
+++ b/src/network/networksocket.c
@@ -4,6 +4,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <libubox/blobmsg_json.h>
+#include <pthread.h>
 
 #include "networksocket.h"
 #include "datastorage.h"
