--- a/block.c
+++ b/block.c
@@ -42,7 +42,7 @@
 #include <libubox/blobmsg_json.h>
 #include <libubox/list.h>
 #include <libubox/ulog.h>
-#include <libubox/utils.h>
+#include "mkdir_p.h"
 #include <libubox/vlist.h>
 #include <libubus.h>
 
--- a/blockd.c
+++ b/blockd.c
@@ -21,6 +21,7 @@
 #include <libubus.h>
 
 #include "libfstools/libfstools.h"
+#include "mkdir_p.h"
 
 #define	AUTOFS_MOUNT_PATH	"/tmp/run/blockd/"
 #define AUTOFS_TIMEOUT		30
--- /dev/null
+++ b/mkdir_p.h
@@ -0,0 +1,40 @@
+
+#ifndef _MKDIR_P_H_
+#define _MKDIR_P_H_
+
+#include <errno.h>
+#include <string.h>
+#include <sys/stat.h>
+
+static int mkdir_p(char *dir, mode_t mask)
+{
+       char *l;
+       int ret;
+
+       ret = mkdir(dir, mask);
+       if (!ret || (ret && errno == EEXIST))
+               return 0;
+
+       if (ret && (errno != ENOENT))
+               return -1;
+
+       l = strrchr(dir, '/');
+       if (!l || l == dir)
+               return -1;
+
+       *l = '\0';
+
+       if (mkdir_p(dir, mask))
+               return -1;
+
+       *l = '/';
+
+       ret = mkdir(dir, mask);
+       if (!ret || (ret && errno == EEXIST))
+               return 0;
+       else
+               return -1;
+}
+
+#endif
+
