--- a/block.c
+++ b/block.c
@@ -767,6 +767,7 @@ static void check_filesystem(struct prob
 	const char *fatfsck = "/usr/sbin/fsck.fat";
 	const char *btrfsck = "/usr/bin/btrfsck";
 	const char *ntfsck = "/usr/bin/ntfsfix";
+	const char *exfatfsck = "/usr/sbin/fsck.exfat";
 	const char *ckfs;
 
 	/* UBIFS does not need stuff like fsck */
@@ -783,6 +784,8 @@ static void check_filesystem(struct prob
 		ckfs = btrfsck;
 	} else if (!strncmp(pr->type, "ntfs", 4)) {
 		ckfs = ntfsck;
+	} else if (!strncmp(pr->type, "exfat", 5)) {
+		ckfs = exfatfsck;
 	} else {
 		ULOG_ERR("check_filesystem: %s is not supported\n", pr->type);
 		return;
