diff --git a/drivers/net/wireless/ath/ath9k/gpio.c b/drivers/net/wireless/ath/ath9k/gpio.c
index 37a796b..7bb2e1f 100644
--- a/drivers/net/wireless/ath/ath9k/gpio.c
+++ b/drivers/net/wireless/ath/ath9k/gpio.c
@@ -390,8 +390,11 @@ void ath_init_leds(struct ath_softc *sc)
 	if (sc->sc_ah->led_pin < 0)
 		return;
 
-	snprintf(led_name, sizeof(led_name), "ath9k-%s",
-		 wiphy_name(sc->hw->wiphy));
+	if (pdata && pdata->led_name)
+		strncpy(led_name, pdata->led_name, sizeof(led_name));
+	else
+		snprintf(led_name, sizeof(led_name), "ath9k-%s",
+			 wiphy_name(sc->hw->wiphy));
 
 	if (ath9k_led_blink)
 		trigger = sc->led_default_trigger;
diff --git a/include/linux/ath9k_platform.h b/include/linux/ath9k_platform.h
index f1f2ad4..558445f 100644
--- a/include/linux/ath9k_platform.h
+++ b/include/linux/ath9k_platform.h
@@ -49,6 +49,7 @@ struct ath9k_platform_data {
 
 	int num_leds;
 	const struct gpio_led *leds;
+	const char *led_name;
 
 	unsigned num_btns;
 	const struct gpio_keys_button *btns;
