--- a/mt7603/dma.c
+++ b/mt7603/dma.c
@@ -195,13 +195,13 @@ int mt7603_dma_init(struct mt7603_dev *dev)
 	for (i = 0; i < ARRAY_SIZE(wmm_queue_map); i++) {
 		ret = mt7603_init_tx_queue(dev, &dev->mt76.q_tx[i],
 					   wmm_queue_map[i],
-					   MT_TX_RING_SIZE);
+					   MT7603_TX_RING_SIZE);
 		if (ret)
 			return ret;
 	}
 
 	ret = mt7603_init_tx_queue(dev, &dev->mt76.q_tx[MT_TXQ_PSD],
-				   MT_TX_HW_QUEUE_MGMT, MT_TX_RING_SIZE);
+				   MT_TX_HW_QUEUE_MGMT, MT7603_PSD_RING_SIZE);
 	if (ret)
 		return ret;
 
--- a/mt7603/mt7603.h
+++ b/mt7603/mt7603.h
@@ -17,6 +17,8 @@
 
 #define MT7603_MCU_RX_RING_SIZE	64
 #define MT7603_RX_RING_SIZE     128
+#define MT7603_TX_RING_SIZE	256
+#define MT7603_PSD_RING_SIZE	128
 
 #define MT7603_FIRMWARE_E1	"mt7603_e1.bin"
 #define MT7603_FIRMWARE_E2	"mt7603_e2.bin"
