From d582e0aba60110f8584adb5cce78ae643dd53e7c Mon Sep 17 00:00:00 2001
From: HiGarfield <HiGarfield@126.com>
Date: Thu, 13 May 2021 17:30:24 +0800
Subject: [PATCH] fix possible NULL pointer dereference in mt76_tx

---
 tx.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tx.c b/tx.c
index d8252e2..e26ea0e 100644
--- a/tx.c
+++ b/tx.c
@@ -294,7 +294,7 @@ mt76_tx(struct mt76_phy *phy, struct ieee80211_sta *sta,
 		skb_set_queue_mapping(skb, qid);
 	}
 
-	if (!(wcid->tx_info & MT_WCID_TX_INFO_SET))
+	if (wcid && !(wcid->tx_info & MT_WCID_TX_INFO_SET))
 		ieee80211_get_tx_rates(info->control.vif, sta, skb,
 				       info->control.rates, 1);
 
-- 
2.17.1

