diff --git a/dma.c b/dma.c
index dbf006a..5f4a1b7 100644
--- a/dma.c
+++ b/dma.c
@@ -563,8 +563,8 @@ mt76_dma_init(struct mt76_dev *dev)
 	init_dummy_netdev(&dev->napi_dev);
 
 	mt76_for_each_q_rx(dev, i) {
-		netif_threaded_napi_add(&dev->napi_dev, &dev->napi[i],
-					mt76_dma_rx_poll, 64);
+		netif_napi_add(&dev->napi_dev, &dev->napi[i], mt76_dma_rx_poll,
+			       64);
 		mt76_dma_rx_fill(dev, &dev->q_rx[i]);
 		napi_enable(&dev->napi[i]);
 	}
