From 77d5bf05d0bb04ec233f81d0cd975ddeaad30dbf Mon Sep 17 00:00:00 2001
From: Emmanuel Engelhart <kelson@kiwix.org>
Date: Sat, 5 Mar 2022 19:15:48 +0100
Subject: [PATCH] Increase ByteArrayDiskWriter maximum size

---
 src/ByteArrayDiskWriter.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ByteArrayDiskWriter.h b/src/ByteArrayDiskWriter.h
index a52cb8c..0e8efb4 100644
--- a/src/ByteArrayDiskWriter.h
+++ b/src/ByteArrayDiskWriter.h
@@ -49,7 +49,7 @@ private:
   void clear();
 
 public:
-  ByteArrayDiskWriter(size_t maxLength = 5_m);
+  ByteArrayDiskWriter(size_t maxLength = 15_m);
   virtual ~ByteArrayDiskWriter();
 
   virtual void initAndOpenFile(int64_t totalLength = 0) CXX11_OVERRIDE;
-- 
2.25.1

