From 63d5d55758c23c04ca8f489c6db81fb91907f2f0 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Sun, 3 Apr 2022 18:53:56 +0900
Subject: [PATCH] Fix a failed test

---
 test/BittorrentHelperTest.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/test/BittorrentHelperTest.cc b/test/BittorrentHelperTest.cc
index 2ad4480..4b3f064 100644
--- a/test/BittorrentHelperTest.cc
+++ b/test/BittorrentHelperTest.cc
@@ -329,6 +329,7 @@ void BittorrentHelperTest::testGetPeerId()
 
 void BittorrentHelperTest::testGetPeerAgent()
 {
+  setStaticPeerAgent("");
   std::string peerAgent = generateStaticPeerAgent("aria2/-1.-1.-1");
   CPPUNIT_ASSERT_EQUAL(std::string("aria2/-1.-1.-1"), peerAgent);
   CPPUNIT_ASSERT_EQUAL(std::string("aria2/-1.-1.-1"),
-- 
2.25.1

