From 851a74f53e6128019dbfff7d410e91c9d4d03799 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Sat, 9 Apr 2022 15:50:42 +0900
Subject: [PATCH 14/23] Fix configure

---
 configure.ac | 1 -
 1 file changed, 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index bae3fce8..56481fd2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -660,7 +660,6 @@ esac
 
 # Checks for header files.
 AC_FUNC_ALLOCA
-AC_CHECK_INCLUDES_DEFAULT
 AC_PROG_EGREP
 
 AC_CHECK_HEADERS([argz.h \
-- 
2.25.1

