From b60a80f1a222a4394978d3ae2a53adf8bb8e4b8f Mon Sep 17 00:00:00 2001
From: Hernan Martinez <hernan.c.martinez@gmail.com>
Date: Thu, 23 Jun 2022 00:00:25 -0600
Subject: [PATCH] Add missing <array>

---
 src/WinTLSSession.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/WinTLSSession.h b/src/WinTLSSession.h
index 21c7462..14e9bee 100644
--- a/src/WinTLSSession.h
+++ b/src/WinTLSSession.h
@@ -37,7 +37,7 @@
 #define WIN_TLS_SESSION_H
 
 #include <vector>
-
+#include <array>
 #include "common.h"
 #include "TLSSession.h"
 #include "WinTLSContext.h"
-- 
2.25.1

