From c99facadd3f25088fea3cba093557319b32b1aae Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Fri, 15 Sep 2023 21:39:33 +0900
Subject: [PATCH] Fix static link failure against libssh2

---
 configure.ac | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure.ac b/configure.ac
index 10acaba..485fe82 100644
--- a/configure.ac
+++ b/configure.ac
@@ -489,6 +489,10 @@ if test "x$with_libssh2" = "xyes"; then
   PKG_CHECK_MODULES([LIBSSH2], [libssh2], [have_libssh2=yes], [have_libssh2=no])
   if test "x$have_libssh2" = "xyes"; then
     AC_DEFINE([HAVE_LIBSSH2], [1], [Define to 1 if you have libssh2.])
+
+    if test "x$ARIA2_STATIC" = "xyes"; then
+      LIBSSH2_CFLAGS="-DLIBSSH2_API= $LIBSSH2_CFLAGS"
+    fi
   else
     AC_MSG_WARN([$LIBSSH2_PKG_ERRORS])
     if test "x$with_libssh2_requested" = "xyes"; then
-- 
2.25.1

