From 69802c058cc0f169e6bb50739fcb661643492c4f Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Sat, 16 Sep 2023 19:17:40 +0900
Subject: [PATCH] Fix overflow

---
 src/cookie_helper.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cookie_helper.cc b/src/cookie_helper.cc
index 8ad1e51..18540c6 100644
--- a/src/cookie_helper.cc
+++ b/src/cookie_helper.cc
@@ -302,7 +302,7 @@ std::unique_ptr<Cookie> parse(const std::string& cookieStr,
           int64_t n = creationTime;
 
           if (n > std::numeric_limits<int64_t>::max() - delta) {
-            maxAge = std::numeric_limits<int64_t>::max();
+            maxAge = std::numeric_limits<time_t>::max();
           } else {
             n += delta;
 
-- 
2.25.1

