From e0a195a48b393b01d5b103fc0a448ae8e2dc37e9 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Mon, 18 Sep 2023 10:22:22 +0900
Subject: [PATCH] Move instead of copy

---
 src/DHTRoutingTableDeserializer.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/DHTRoutingTableDeserializer.cc b/src/DHTRoutingTableDeserializer.cc
index 87b757a..63a82c7 100644
--- a/src/DHTRoutingTableDeserializer.cc
+++ b/src/DHTRoutingTableDeserializer.cc
@@ -196,8 +196,8 @@ void DHTRoutingTableDeserializer::deserialize(const std::string& filename)
 
     nodes.push_back(node);
   }
-  localNode_ = localNode;
-  nodes_ = nodes;
+  localNode_ = std::move(localNode);
+  nodes_ = std::move(nodes);
   A2_LOG_INFO("DHT routing table was loaded successfully");
 }
 
-- 
2.25.1

