From 84c1e0ad32315e2814b6fe474a2fb21045c538e3 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Mon, 18 Sep 2023 11:00:00 +0900
Subject: [PATCH] Bail out if pieceLength is 0

---
 src/DefaultBtProgressInfoFile.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/DefaultBtProgressInfoFile.cc b/src/DefaultBtProgressInfoFile.cc
index 7a5da1c..b216182 100644
--- a/src/DefaultBtProgressInfoFile.cc
+++ b/src/DefaultBtProgressInfoFile.cc
@@ -289,6 +289,10 @@ void DefaultBtProgressInfoFile::load()
     pieceLength = ntohl(pieceLength);
   }
 
+  if (pieceLength == 0) {
+    throw DL_ABORT_EX("piece length must not be 0");
+  }
+
   uint64_t totalLength;
   READ_CHECK(fp, &totalLength, sizeof(totalLength));
   if (version >= 1) {
-- 
2.25.1

