From 3412fa3513f8f4ac93cd6c1c730bcc5ffd37c02c Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Mon, 18 Sep 2023 17:13:09 +0900
Subject: [PATCH] Do not close stdout and stderr

---
 src/BufferedFile.cc | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/BufferedFile.cc b/src/BufferedFile.cc
index 0f9a97b..fe1b977 100644
--- a/src/BufferedFile.cc
+++ b/src/BufferedFile.cc
@@ -86,7 +86,9 @@ int BufferedFile::onClose()
 #else  // __MINGW32__
     _commit(fileno(fp_));
 #endif // __MINGW32__
-    rv = fclose(fp_);
+    if (fp_ != stdin && fp_ != stderr) {
+      rv = fclose(fp_);
+    }
     fp_ = nullptr;
   }
   return rv;
-- 
2.25.1

