From 94725a037129901fa81fb14e56be529ccbeed9e7 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Fri, 3 Nov 2023 10:51:51 +0900
Subject: [PATCH 56/57] Fix build error without bittorrent

Amends #2127
---
 src/DefaultBtProgressInfoFile.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/DefaultBtProgressInfoFile.cc b/src/DefaultBtProgressInfoFile.cc
index 9d03e3e9..7907284f 100644
--- a/src/DefaultBtProgressInfoFile.cc
+++ b/src/DefaultBtProgressInfoFile.cc
@@ -57,6 +57,7 @@
 #include "DownloadContext.h"
 #include "BufferedFile.h"
 #include "SHA1IOFile.h"
+#include "BtConstants.h"
 #ifdef ENABLE_BITTORRENT
 #  include "PeerStorage.h"
 #  include "BtRuntime.h"
-- 
2.25.1

